junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/discovery/LauncherDiscoveryListeners.java

```diff
@@ -30,12 +38,50 @@ public static LauncherDiscoveryListener logging() {
 		return new LoggingLauncherDiscoveryListener();
 	}
 
+	@API(status = API.Status.INTERNAL, since = "1.6")
 	public static LauncherDiscoveryListener composite(List<LauncherDiscoveryListener> listeners) {
 		Preconditions.notEmpty(listeners, "listeners must not be empty");
+		Preconditions.containsNoNullElements(listeners, "listeners must not contain any null elements");
 		if (listeners.size() == 1) {
 			return listeners.get(0);
 		}
 		return new CompositeLauncherDiscoveryListener(listeners);
 	}
 
+	@API(status = API.Status.INTERNAL, since = "1.6")
+	public static LauncherDiscoveryListener fromConfigurationParameter(
+			ConfigurationParameters configurationParameters) {
+		return configurationParameters.get(DEFAULT_DISCOVERY_LISTENER_CONFIGURATION_PROPERTY_NAME,
+			defaultListenerType -> //
+			Arrays.stream(LauncherDiscoveryListenerType.values()) //
+					.filter(type -> type.parameterValue.equalsIgnoreCase(defaultListenerType)) //
+					.findFirst() //
+					.map(type -> type.creator.get()) //
+					.orElseThrow(() -> {
+						String allowedValues = Arrays.stream(LauncherDiscoveryListenerType.values()) //
+								.map(type -> type.parameterValue) //
+								.collect(joining("', '", "'", "'"));
+						return new JUnitException("Invalid value of configuration parameter '"
+								+ DEFAULT_DISCOVERY_LISTENER_CONFIGURATION_PROPERTY_NAME + "': " //
+								+ defaultListenerType + " (allowed are " + allowedValues + ")");
+					})) //
+				.orElseGet(LauncherDiscoveryListeners::abortOnFailure);
+	}
```

Comment:
A rather deeply nested indentation - Did you consider a temp variable or helper method to reduce indentation?

Reply:
I didn't look quite as bad in an (my) IDE but I extracted a method nevertheless.